contracts/PuppySponsor.sol

```diff
@@ -0,0 +1,193 @@
+    EnumerableSet.Bytes32Set private _puppyIdSet;
+
+    // An unique identifier that represents not a puppy.
+    bytes32 private constant NOT_PUPPY_ID = bytes32(0);
+
+    // The empty string.
+    string private constant EMPTY_STRING = "";
+
+    event Donated(address indexed from, address indexed dest, bytes32 puppyId, uint256 amount, uint256 time);
+    event PuppyAdded(bytes32 indexed puppyId, string name, string birthday, string imageUrl, string description, uint256 time);
+
+    // A donation transaction, may for a specific puppy.
+    struct DonateTransaction {
+        // The EOA address of the transaction initiator, i.e. the donor.
+        address donor;
+        // The EOA address of the transaction payee.
+        address receiver;
+        // The amount of the transaction, in (wei).
+        uint256 amount;
+        // The block time when the transaction was on-chain.
+        uint256 time;
+        // The puppy's unique identification code.
+        // But if the transaction is not for a specific puppy (donations for feed), the value is NOT_PUPPY_ID.
+        bytes32 puppyId;
+        // Information attached by the donor to this donation.
+        string message;
+        // Donor-customized keywords.
+        string keyword;
+    }
+
+    // A puppy entity, which contains its information.
+    struct Puppy {
+        // The puppy's unique identification code.
+        bytes32 puppyId;
+        // The puppy's name.
+        string name;
+        // The puppy's birthday.
+        string birthday;
+        // An image's url for the puppy.
+        string imageUrl;
+        // The detail information for the puppy.
+        string description;
+    }
+
+    /**
+     * @dev Compare if the provided string is same.
+     * Using keccak256 hash to compare.
+     */
+    function _strEq(string memory a, string memory b) internal pure returns (bool) {
+        if (bytes(a).length != bytes(b).length) {
+            return false;
+        }
+
+        return keccak256(abi.encodePacked(a)) == keccak256(abi.encodePacked(b));
+    }
+
+    /**
+     * @dev Check if puppyId is valid.
+     */
+    function _isValidPuppyId(bytes32 puppyId) private view returns (bool) {
+        return _puppyIdSet.contains(puppyId);
+    }
+
+    /**
+     * @dev donate for puppy or food.
+     * Requirements:
+     *
+     * - `msg.value` must be greater than zero.
+     *
+     * Emits a {Donated} event to record the new donation.
+     *
+     */
+    function _donate(bytes32 puppyId, string calldata message, string calldata keyword) private {
+        require(msg.value > 0, "The transaction value must be greater than zero!");
+
+        donateTransactionAmount++;
+        totalReceivedDonation += msg.value;
+
+        payable(owner()).transfer(msg.value);
+
+        DonateTransaction memory newTransaction = DonateTransaction(msg.sender, owner(), msg.value, block.timestamp, puppyId, message, keyword);
+        donateTransactions.push(newTransaction);
```

Comment:
Why don't we write it inline?

```suggestion
        donateTransactions.push(DonateTransaction(msg.sender, owner(), msg.value, block.timestamp, puppyId, message, keyword));
```

Reply:
In fact, I think that's fine too.
The reason why it was considered to be dismantled before was that it was a little long and looked a little protruding, so it was dismantled.

Reply:
Oh I see, then I think it's fine to keep it.